concurrency/impl-31/src/main/java/org/jboss/as/ee/concurrent/ConcurroManagedThreadFactoryImpl.java

```diff
@@ -0,0 +1,129 @@
+ */
+
+package org.jboss.as.ee.concurrent;
+
+import org.glassfish.concurro.AbstractManagedThread;
+import org.glassfish.concurro.ContextServiceImpl;
+import org.glassfish.concurro.internal.ManagedFutureTask;
+import org.glassfish.concurro.spi.ContextHandle;
+import org.wildfly.security.auth.server.SecurityIdentity;
+import org.wildfly.security.manager.WildFlySecurityManager;
+
+import jakarta.enterprise.concurrent.ManagedThreadFactory;
+import java.security.AccessControlContext;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+
+/**
+ * {@link ManagedThreadFactory} implementation ensuring {@link SecurityIdentity} propagation into new threads.
+ * @author <a href="mailto:jkalina@redhat.com">Jan Kalina</a>
+ * @author emmartins
+ */
+public class ConcurroManagedThreadFactoryImpl extends org.glassfish.concurro.ManagedThreadFactoryImpl implements WildFlyManagedThreadFactory {
+
+    /**
+     * the priority set on new threads
+     */
+    private final int priority;
+
+    /**
+     * the factory's ACC
+     */
+    private final AccessControlContext accessControlContext;
+
+    public ConcurroManagedThreadFactoryImpl(String name, WildFlyContextService contextService, int priority) {
+        super(name, (ContextServiceImpl) contextService, priority);
+        this.priority = priority;
+        this.accessControlContext = AccessController.getContext();
+    }
+
+    /**
+     *
+     * @return the priority set on new threads
+     */
+    public int getPriority() {
+        return priority;
+    }
+
+    protected AbstractManagedThread createThread(Runnable r, final ContextHandle contextHandleForSetup) {
+        if (contextHandleForSetup != null) {
+            // app thread, do identity wrap
+            r = SecurityIdentityUtils.doIdentityWrap(r);
+        }
+        // use the factory's acc as privileged, otherwise the new thread inherits current thread's acc
+        final AbstractManagedThread t = AccessController.doPrivileged(new CreateThreadAction(r, contextHandleForSetup), accessControlContext);
+        // reset thread classloader to prevent leaks
+        if (!WildFlySecurityManager.isChecking()) {
+            t.setContextClassLoader(null);
+        } else {
+            AccessController.doPrivileged((PrivilegedAction<Void>) () -> {
+                t.setContextClassLoader(null);
+                return null;
+            });
+        }
+        return t;
+    }
+
+    @Override
+    public void taskStarting(Thread t, ManagedFutureTask task) {
+        super.taskStarting(t, task);
+        if (t instanceof ManagedThread) {
+            ((ManagedThread)t).task = task;
+        }
+    }
+
+    @Override
+    public void taskDone(Thread t) {
+        super.taskDone(t);
+        if (t instanceof ManagedThread) {
+            ((ManagedThread)t).task = null;
+        }
+    }
+
+    private final class CreateThreadAction implements PrivilegedAction<AbstractManagedThread> {
```

Comment:
This should no longer be needed as the security manager support is removed from Jakarta EE 11.